api/src/server/gql/mutation.rs

```diff
@@ -25,6 +28,73 @@ use validator::Validate;
 pub struct Mutation;
 
 impl MutationFields for Mutation {
+    fn field_initialize_user(
+        &self,
+        executor: &juniper::Executor<'_, Context>,
+        _trail: &QueryTrail<'_, InitializeUserPayload, Walked>,
+        input: InitializeUserInput,
+    ) -> ResponseResult<InitializeUserPayload> {
+        let context = executor.context();
+
+        match context.user_context {
+            Some(UserContext {
+                user_provider_id, ..
+            }) => {
+                let conn = &context.get_db_conn()?;
+                let new_user = models::NewUser {
+                    username: &input.username,
+                };
+                new_user.validate()?;
+
+                // We need to insert the new user row, then update the
+                // user_provider to point at that row. We need a transaction to
+                // prevent race conditions.
+                let created_user = conn
+                    .transaction::<models::User, ResponseError, _>(|| {
+                        let create_user_result: Result<models::User, _> =
+                            new_user
+                                .insert()
+                                .returning(users::all_columns)
+                                .get_result(conn);
+
+                        // Check if the username already exists
+                        let created_user = DbErrorConverter {
+                            unique_violation_to_exists: true,
+                            ..Default::default()
+                        }
+                        .convert(create_user_result)?;
+
+                        // We should update exactly 1 row. If not, then either
+                        // the referenced user_provider row is already linked to
+                        // a user, or it doesn't exist. In either case, just
+                        // return a NotFound error.
+                        let updated_rows = diesel::update(
+                            user_providers::table
+                                .find(user_provider_id)
+                                .filter(
+                                    user_providers::columns::user_id.is_null(),
+                                ),
+                        )
+                        .set(
+                            user_providers::columns::user_id
+                                .eq(Some(created_user.id)),
+                        )
+                        .execute(conn)?;
+
+                        if updated_rows == 0 {
+                            Err(ResponseError::NotFound)
+                        } else {
+                            Ok(created_user)
+                        }
+                    })?;
```

Comment:
Can you make sure this transaction logic makes sense? I doubled checked in the diesel source, and if this lambda returns any `Err` it will roll back. I also tried to verify that in the test. We should be good here if there are simultaneous requests, right? One should go through properly while the other reverts.

Reply:
> We should be good here if there are simultaneous requests, right? 

What would be simultaneous, isn't it synchronous since it updates the `users` table then `user_providers`?

Reply:
If someone sent the same mutation request twice simultaneously, we could potentially have multiple threads in this critical section at the same time.

Reply:
I think it would error out correctly on commit in that case but tbh im not sure. We can just leave it as a thing to look into since i doubt it would ever actually happen